### PR TITLE
Feature: finetuning

### PR DESCRIPTION
**Pull request by sgsdxzy**
This would allow to use different caption for each image, allowing:

1. better regularization
2. train multiple concepts more easily
3. finetune the entire model

Implements #72 

Roadmap:

- [ ] read captions from directories prepared by webui preprocess
- [ ] merge instance prompt to caaptions
- [ ] ema weights for model finetuning

### CODE REVIEW

**Comment by sgsdxzy** [comment]
@d8ahazard need your idea on the tokenizer part:
Currently we use the tokenizer from checkpoint directly, and it does not support the fancy part webui extends, like more than 77 token length, which is crucial for supporting reading long captions. It is relatively easy to at least support unlimited length, but I don't know whether it would interfere with tokenizer training. Do we turn off tokenizer training when user opt to use unlimited token length?

**Comment by d8ahazard** [comment]
That is a great question, let me think on it a bit (I just woke up).

My question is - this looks like the same-ish feature as #92, which is great. I just need to know who's work to use and who to credit? :P

**Comment by sgsdxzy** [comment]
> That is a great question, let me think on it a bit (I just woke up).
> 
> My question is - this looks like the same-ish feature as #92, which is great. I just need to know who's work to use and who to credit? :P

#92 looks good enough. I plan to use the wrapping provided by webui it self, and allow user to write something like `zwgs dog, {filewords}` as instance prompt to easily integrate captions, I think it is more versatile but I still need some time.

#92 can already work on its own, and not complicated.

**Comment by sgsdxzy** [comment]
I am still trying to get advanced tokenizer features to work. For example, user may create deepdanbooru tags of franchise name like `le malin \(azur lane\)`, without advance parsing the tokenizer cannot handle the `\(`.

**Comment by dali99** [comment]
How are you planning on handling classification images with this? I tried finetuning a model with descriptions earlier but when words in the captions showed up a lot they become "part" of the prompt.

Specifically I would have to write zkz looking at viewer to get what would normally be just zkz. I'm guessing this has to to with those parts of the captions not getting any regularization done on them so they are much more prone to overfitting.

One solutution would be to generate regularization images which keep only the [filewords]. But you know a lot more about this than I do so I wanted to hear your thoughts

**Comment by sgsdxzy** [comment]
> How are you planning on handling classification images with this? I tried finetuning a model with descriptions earlier but when words in the captions showed up a lot they become "part" of the prompt.
> 
> Specifically I would have to write zkz looking at viewer to get what would normally be just zkz. I'm guessing this has to to with those parts of the captions not getting any regularization done on them so they are much more prone to overfitting.
> 
> One solution would be to generate regularization images which keep only the [filewords]. But you know a lot more about this than I do so I wanted to hear your thoughts

No, I have no experience at all, I am writing code so that I could finally start training.
As for finetuning, I never thought about using regularization in the first place. I think with enough images it should be fine without regularization or overfitting. But If you are trying to do few sample training, then probably you should go with normal dreambooth, not finetuning mode.
How many samples are you using? I prepared ~2k.

**Comment by d8ahazard** [comment]
I just mentioned this in another discussion, but I'm curious what would happen if we had a wider set of regularization images based on lists of prompts - with the prompts saved in the filename/txt with the class image, and this prompt is used when comparing the class image.

Seems like it could only help...

**Comment by sgsdxzy** [comment]
On the other hand if you train with 'dog zwx' and class image 'dog', you can use just zwx to invoke but it would be very hard to get a long tail version of zwx.

**Comment by sgsdxzy** [comment]
So probably we should implement the feature like this:
If user gives a class prompt without [filewords], keep it as is;
If user gives a class prompt with [filewords], and a directory with enough number of images, just read the filewords of the class image to be part of the class prompt.
If user gives a class prompt with [filewords], and an empty directory, and a nonzero class image number, we generate that many images, each time with a randomly chosen(or round-robin?) filewords of an instance image. Tag shuffle and dropout from settings also apply.
What do you think @d8ahazard

**Comment by sgsdxzy** [comment]
Update: implemented EMA. Default off so it should not introduce any problem is not explicitly switched on.
However, the ema model cannot fit into 12G VRAM so I am not able to test it...

**Comment by d8ahazard** [comment]
> Update: implemented EMA. Default off so it should not introduce any problem unless is not explicitly switched on. However, the ema model cannot fit into 12G VRAM so I am not able to test it...

I definitely like that everything "extra" is stuck in a separate file. The idea is that the base "train_dreambooth.py" file can be updated as quickly as possible when updates are put to the base repo, so having as much of our new code in a different class file is ideal.

I've got a 24GB GPU, I'll test the EMA training today.

Also, since you seem to be pretty adept at schlepping in "other" bits of tuning code, what do you think about the concept of "aspect-ratio bucketing"?

https://github.com/NovelAI/novelai-aspect-ratio-bucketing

**Comment by sgsdxzy** [comment]
@d8ahazard I think this is ready for review.
The tokenizer part is too complicated and require some major overhual to DreamboothDataset and collate_fn. I don't know torch and cannot get the simple solution from https://github.com/bmaltais/kohya_ss to work.

**Comment by d8ahazard** [comment]
Testing this now. Just so we're clear, if I use [filewords] instead of my instance prompt, it will just read the prompt from the instance image file name, right? 

So then what about class images? If I want to generate new, dynamic class images, how do I do that?

**Comment by d8ahazard** [comment]
Gotcha. So, then, in it's current iteration, I think everything here is fine. 

Testing EMA training now, that also appears to be working on my 24GB GPU.

Only thing I would maybe like to see changed is adding some way to read class prompts from a txt file or something else.

Logic for this is that I've created another extension for dreambooth that allows auto-captioning of instance images, with the ability to swap out the class for the specific subject word. So, my images are already captioned "a photo of ZKS, white shirt, beach", etc. 

Hence, using filewords for my class prompts is a non-starter. Not the end of the world, but in a perfect scenario, we'd have a way to specify like a list of class prompts or swap the instance word with the class word or something.

But, we'll save that for another idea some time down the road...

**Comment by sgsdxzy** [comment]
> Gotcha. So, then, in it's current iteration, I think everything here is fine.
> 
> Testing EMA training now, that also appears to be working on my 24GB GPU.
> 
> Only thing I would maybe like to see changed is adding some way to read class prompts from a txt file or something else.
> 
> Logic for this is that I've created another extension for dreambooth that allows auto-captioning of instance images, with the ability to swap out the class for the specific subject word. So, my images are already captioned "a photo of ZKS, white shirt, beach", etc.
> 
> Hence, using filewords for my class prompts is a non-starter. Not the end of the world, but in a perfect scenario, we'd have a way to specify like a list of class prompts or swap the instance word with the class word or something.
> 
> But, we'll save that for another idea some time down the road...

I understand what you are trying to achieve. Yes, for now the images is better captioned without subject/class name, to be given one by prompt template. On the other hand, do you plan to make class image generation in dreambooth a simple tool, or fully-fledged? I think you can always caption your instance images, use text replace tools to replace the captions, and use the new captions to generate images in some other pipelines, then just read these image and the modified captions back using [filewords] mechanism.

One other thing I'd like to note is that this EMA implementation not complete. To do some serious training, one should start from the 7G model, which contains both full weight and ema weight. One should init the training weights using full weight, init the ema weight using ema weight from ckpt, train in full weight and update ema weight every step or epoch, save both weights on end, and distribute/use ema weight for inference. This is how real training is done. But most users just start from the 4G model and cannot afford to load two weights simutaneously anyway. In this scenario, ema might not be very helpful (but it at least lessens overfit a little).

**Comment by d8ahazard** [comment]
Getting an error when I start training using 8bit/fp16:

line 38 in finetune_utils.py
def create_text(self, text_template, filename_text):
        tags = filename_text.split(',')
        if shared.opts.tag_drop_out != 0:
        
        Can't find shared.opts.tag_drop_out or shared.opts.shuffle_tags:
        
This is on a brand-new clone of the repo and fresh install.

**Comment by sgsdxzy** [comment]
> Getting an error when I start training using 8bit/fp16:
> 
> line 38 in finetune_utils.py def create_text(self, text_template, filename_text): tags = filename_text.split(',') if shared.opts.tag_drop_out != 0:
> 
> ```
>     Can't find shared.opts.tag_drop_out or shared.opts.shuffle_tags:
> ```
> 
> This is on a brand-new clone of the repo and fresh install.

Did you update webui? It requires https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/73776907ec460747f37e1154d890a44fd83b8440 or later.

**Comment by d8ahazard** [comment]
Yep, I'm a big ol' dumbass. I was testing on my 8GB GPU, pulled the wrong branch. Sorry for blowing up.:P

**Comment by d8ahazard** [comment]
Just pushed a commit that puts the ema_unet on the accelerator. Curious to know if you get OOM with it on your GPU now.

**Comment by d8ahazard** [comment]
@sgsdxzy -

So, in one of the updates for diffusers, they added a flag to "extract_ema" weights from a model if it has them, versus "normal" weights.

This gets me thinking "Why can't we extract and save both"? 

image

Instead of our current process, where we are overwriting the normal unet weights with the ema ones, I'm wondering if we can reverse the above process somehow and save both sets of weights, like in the big boy models...

**Comment by sgsdxzy** [comment]
> Instead of our current process, where we are overwriting the normal unet weights with the ema ones, I'm wondering if we can reverse the above process somehow and save both sets of weights, like in the big boy models...

I thought about this a week ago and was planning to support it, which is not very hard. But you said you were doing a major overhaul so I just wait until you finish :>
On the other hand, most users just start from the 4G ema-only model so it will not make much difference, and there are some detailed I think we need to think over:
1. what if the user extract both but train without ema, then resume training with ema
2. what if the user trained ema, saved both, then resume training without ema
